test/test-lint.cpp

```diff
@@ -2811,6 +2811,201 @@ TEST(
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    // {
+    //   eval("var x = 42");
+    // }
+    // x;
+    // x = 10;
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_block_scope();
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_exit_block_scope();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    // eval("var x = 42");
+    // function f() {
+    //   x;
+    //   x = 10;
+    // }
+    error_collector v;
+    linter l(&v);
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_exit_function_scope();
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    // function f() {
+    //   x;
+    //   x = 10;
+    // }
+    // eval("var x = 42");
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_exit_function_scope();
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    // function f() {
+    //   x;
+    //   x = 10;
+    //   eval("var x = 42;");
+    //   x;
+    //   x = 10;
+    // }
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_exit_function_scope();
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    // function f() {
+    //   x;
+    //   x = 10;
+    //   {
+    //     eval("var x = 42;");
+    //   }
+    //   x;
+    //   x = 10;
+    // }
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_enter_block_scope();
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_exit_block_scope();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_exit_function_scope();
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+  {
+    // function f() {
+    //   eval("var x = 42;");
+    // }
+    // function g() {
+    //   x;
```

Comment:
For tests with errors, I like to include a comment pointing out where the error is:
```suggestion
    //   x;  // ERROR
```